compiler/optimizer/VPHandlers.cpp

```diff
@@ -4241,13 +4241,65 @@ TR::Node *constrainNew(OMR::ValuePropagation *vp, TR::Node *node)
          vp->addGlobalConstraint(node, classConstraint->asClass()->getClassType());
       else
          vp->addGlobalConstraint(node, classConstraint);
+
+      // If it's an abstract class or an interface, the allocation cannot be removed
+      // because InstantiationError could be thrown
+      TR_OpaqueClassBlock *clazz = classConstraint->getClassType() ? classConstraint->getClassType()->getClass() : NULL;
+      if (clazz && TR::Compiler->cls.isConcreteClass(vp->comp(), clazz))
+         {
+         node->setAllocationCanBeRemoved(true);
+         }
       }
 
-   if (classConstraint && classConstraint->getClassType() &&
-      classConstraint->getClassType()->getClass())
+   vp->addGlobalConstraint(node, TR::VPNonNullObject::create(vp));
+
+   node->setIsNonNull(true);
+
+   return node;
+   }
+
+TR::Node *constrainNewvalue(OMR::ValuePropagation *vp, TR::Node *node)
+   {
+   constrainChildren(vp, node);
+
+   vp->createExceptionEdgeConstraints(TR::Block::CanCatchNewvalue, NULL, node);
+
+   bool trace = vp->trace();
+
+   // The constraints on the child can be applied to this node. We also know
+   // that the type is fixed and non-null.
+   //
+   bool isGlobal;
+   TR::VPConstraint *classConstraint = vp->getConstraint(node->getFirstChild(), isGlobal);
+   if (classConstraint)
       {
-      node->setAllocationCanBeRemoved(true);
+      // since loadaddrs are now primed with a ClassObject property,
+      // this property should not be propagated to the 'newvalue' unless the
+      // underlying type of the loadaddr is a java/lang/Class
+      //
+      // If the underlying type of the loadaddr is a java/lang/Class, the property can be propagated to 'new'
+      //
+      if (classConstraint->getClass() && !classConstraint->isFixedClass())
+         {
+         vp->addGlobalConstraint(node, TR::VPFixedClass::create(vp, classConstraint->getClass()));
+         }
+      else if (classConstraint->asClass() && classConstraint->asClass()->getClassType() &&
+               !(classConstraint->asClass()->getClassType()->isClassObject() == TR_yes))
+         {
+         vp->addGlobalConstraint(node, classConstraint->asClass()->getClassType());
+         }
+      else
+         {
+         vp->addGlobalConstraint(node, classConstraint);
+         }
+
+      // Primitive value type cannot be abstract class
```

Comment:
This should say `// Value type cannot be abstract class`.  The `newvalue` can create instances of any value type classes, not just primitive value type classes.

Reply:
Addressed in f2616d4